File: _정답코드/4주차/04_쿼드트리_이미지_압축/solution.cpp
#include <vector>
#include <string>
using namespace std;

vector<vector<int>> imgGlob;

string encode(int x, int y, int n) {
    int countZero = 0;
    for (int i = x; i < x + n; i++) {
        for (int j = y; j < y + n; j++) {
            if (imgGlob[i][j] == 0) {
                countZero++;
            }
        }
    }
    if (countZero == 0) {
        return "1";
    }
    if (countZero == n * n) {
        return "0";
    }
    return "("
        + encode(x, y, n / 2)
        + encode(x, y + n / 2, n / 2)
        + encode(x + n / 2, y, n / 2)
        + encode(x + n / 2, y + n / 2, n / 2)
        + ")";
}

string solution(vector<vector<int>> img) {
    imgGlob = img;
    return encode(0, 0, 8);
}

File: _정답코드/4주차/04_쿼드트리_이미지_압축/solution.py
img_glob = []

def solution(img):
    global img_glob
    img_glob = img
    return encode(0, 0, 8)

def encode(x, y, n):
    count_zero = 0
    for i in range(x, x + n):
        for j in range(y, y + n):
            if img_glob[i][j] == 0:
                count_zero += 1
    if count_zero == 0:
        return "1"
    if count_zero == n * n:
        return "0"
    return "(" + \
        encode(x, y, n // 2) + \
        encode(x, y + n // 2, n // 2) + \
        encode(x + n // 2, y, n // 2) + \
        encode(x + n // 2, y + n // 2, n // 2) + \
        ")"

